File: server/mechanism.js
const GENERATION_FACTORS = Object.freeze({ standard: 1, pro: 1.5 });
const CLIP_GRANT_PER_CREDIT = 0.4;
const CLIP_COST_PER_CREDIT = 0.2;
const HAPW_EXCHANGE_FEE_RATE = 0.05;
const HAPW_EXCHANGE_DAILY_LIMIT = 2;

function generationCost(duration, quality) {
  const seconds = Number(duration);
  const factor = GENERATION_FACTORS[quality];
  if (!Number.isFinite(seconds) || seconds <= 0 || !factor) throw new RangeError('invalid_generation_parameters');
  return Math.ceil(seconds * factor);
}

function redemptionClipGrant(credits) {
  const value = Number(credits);
  if (!Number.isFinite(value) || value <= 0) throw new RangeError('invalid_credit_grant');
  return Math.floor(value * CLIP_GRANT_PER_CREDIT);
}

function generationClipCost(credits) {
  const value = Number(credits);
  if (!Number.isFinite(value) || value <= 0) throw new RangeError('invalid_credit_cost');
  return Math.ceil(value * CLIP_COST_PER_CREDIT);
}

function hapwExchangeQuote(price, feeRate = HAPW_EXCHANGE_FEE_RATE) {
  const value = Number(price);
  const rate = Number(feeRate);
  if (!Number.isFinite(value) || value <= 0 || !Number.isFinite(rate) || rate < 0 || rate > 1) throw new RangeError('invalid_exchange_quote');
  const fee = Math.ceil(value * rate);
  return { price: value, fee, total: value + fee, feeRate: rate };
}

function hapwExchangeDailySnapshot(exchanges, dailyLimit = HAPW_EXCHANGE_DAILY_LIMIT, now = new Date()) {
  if (!Array.isArray(exchanges)) throw new TypeError('invalid_exchange_records');
  const limit = Number(dailyLimit);
  if (!Number.isInteger(limit) || limit <= 0) throw new RangeError('invalid_daily_exchange_limit');
  const date = new Date(now);
  if (Number.isNaN(date.getTime())) throw new RangeError('invalid_exchange_date');
  const day = date.toISOString().slice(0, 10);
  const reset = new Date(`${day}T00:00:00.000Z`);
  reset.setUTCDate(reset.getUTCDate() + 1);
  const usedToday = exchanges.filter(item => item && item.statusCode === 'completed' && String(item.createdAt || '').slice(0, 10) === day).length;
  return { date: day, timezone: 'UTC', resetsAt: reset.toISOString(), dailyLimit: limit, usedToday, remainingToday: Math.max(0, limit - usedToday), reached: usedToday >= limit };
}

function dexPoolSnapshot(pool) {
  const clipReserve = Number(pool.clipReserve);
  const usdtReserve = Number(pool.usdtReserve);
  if (!Number.isFinite(clipReserve) || clipReserve <= 0 || !Number.isFinite(usdtReserve) || usdtReserve <= 0) throw new RangeError('invalid_liquidity_pool');
  return {
    clipReserve,
    usdtReserve,
    clipPerUsdt: clipReserve / usdtReserve,
    usdtPerClip: usdtReserve / clipReserve,
    totalLiquidityUsdt: usdtReserve * 2,
    updatedAt: pool.updatedAt
  };
}

function syncGenerationAccount(account, redemptions, generations) {
  account.balance = redemptions
    .filter(item => item.status === '有效')
    .reduce((sum, item) => sum + item.creditsRemaining, 0);
  account.lifetimeGranted = redemptions.reduce((sum, item) => sum + item.creditsGranted, 0);
  account.lifetimeUsed = generations.reduce((sum, item) => sum + item.creditsUsed, 0);
  account.lifetimeClipGranted = redemptions.reduce((sum, item) => sum + (item.clipGranted || 0), 0);
  account.lifetimeClipSpent = generations.reduce((sum, item) => sum + (item.clipCost || 0), 0);
  return account;
}

module.exports = {
  GENERATION_FACTORS,
  CLIP_GRANT_PER_CREDIT,
  CLIP_COST_PER_CREDIT,
  HAPW_EXCHANGE_FEE_RATE,
  HAPW_EXCHANGE_DAILY_LIMIT,
  generationCost,
  redemptionClipGrant,
  generationClipCost,
  hapwExchangeQuote,
  hapwExchangeDailySnapshot,
  dexPoolSnapshot,
  syncGenerationAccount
};

File: internal/service/mechanism.go
package service

import (
	"errors"
	"math"
	"time"

	"github.com/StevenWXY/HAPE/internal/domain"
)

const (
	CLIPGrantPerCredit     = 0.4
	CLIPCostPerCredit      = 0.2
	HAPWExchangeFeeRate    = 0.05
	HAPWExchangeDailyLimit = 2
)

var generationFactors = map[string]float64{"standard": 1, "pro": 1.5}

func GenerationCost(duration int, quality string) (int, error) {
	factor, ok := generationFactors[quality]
	if duration <= 0 || !ok {
		return 0, errors.New("invalid_generation_parameters")
	}
	return int(math.Ceil(float64(duration) * factor)), nil
}

func RedemptionCLIPGrant(credits int) (int, error) {
	if credits <= 0 {
		return 0, errors.New("invalid_credit_grant")
	}
	return int(math.Floor(float64(credits) * CLIPGrantPerCredit)), nil
}

func GenerationCLIPCost(credits int) (int, error) {
	if credits <= 0 {
		return 0, errors.New("invalid_credit_cost")
	}
	return int(math.Ceil(float64(credits) * CLIPCostPerCredit)), nil
}

type ExchangeQuote struct {
	Price   int     `json:"price"`
	Fee     int     `json:"fee"`
	Total   int     `json:"total"`
	FeeRate float64 `json:"feeRate"`
}

func QuoteHAPWExchange(price int, feeRate float64) (ExchangeQuote, error) {
	if price <= 0 || feeRate < 0 || feeRate > 1 {
		return ExchangeQuote{}, errors.New("invalid_exchange_quote")
	}
	fee := int(math.Ceil(float64(price) * feeRate))
	return ExchangeQuote{Price: price, Fee: fee, Total: price + fee, FeeRate: feeRate}, nil
}

type ExchangePolicy struct {
	Date               string `json:"date"`
	Timezone           string `json:"timezone"`
	ResetsAt           string `json:"resetsAt"`
	DailyLimit         int    `json:"dailyLimit"`
	UsedToday          int    `json:"usedToday"`
	RemainingToday     int    `json:"remainingToday"`
	Reached            bool   `json:"reached"`
	InventoryTotal     int    `json:"inventoryTotal"`
	InventoryAvailable int    `json:"inventoryAvailable"`
}

func BuildExchangePolicy(exchanges []domain.HAPWExchange, assets []domain.HAPWAsset, now time.Time) ExchangePolicy {
	day := now.UTC().Format("2006-01-02")
	used := 0
	for _, exchange := range exchanges {
		if exchange.StatusCode == "completed" && len(exchange.CreatedAt) >= 10 && exchange.CreatedAt[:10] == day {
			used++
		}
	}
	total, available := 0, 0
	for _, asset := range assets {
		if asset.Kind == "HAPW" && asset.ClipPrice > 0 {
			total++
			if asset.ExchangeAvailable {
				available++
			}
		}
	}
	remaining := HAPWExchangeDailyLimit - used
	if remaining < 0 {
		remaining = 0
	}
	reset := time.Date(now.UTC().Year(), now.UTC().Month(), now.UTC().Day()+1, 0, 0, 0, 0, time.UTC)
	return ExchangePolicy{
		Date: day, Timezone: "UTC", ResetsAt: reset.Format(time.RFC3339), DailyLimit: HAPWExchangeDailyLimit,
		UsedToday: used, RemainingToday: remaining, Reached: used >= HAPWExchangeDailyLimit,
		InventoryTotal: total, InventoryAvailable: available,
	}
}

type DexPoolSnapshot struct {
	ClipReserve        int     `json:"clipReserve"`
	USDTReserve        int     `json:"usdtReserve"`
	ClipPerUSDT        float64 `json:"clipPerUsdt"`
	USDTPerCLIP        float64 `json:"usdtPerClip"`
	TotalLiquidityUSDT int     `json:"totalLiquidityUsdt"`
	UpdatedAt          string  `json:"updatedAt"`
}

func BuildDexPoolSnapshot(pool domain.CLIPPool) (DexPoolSnapshot, error) {
	if pool.ClipReserve <= 0 || pool.USDTReserve <= 0 {
		return DexPoolSnapshot{}, errors.New("invalid_liquidity_pool")
	}
	return DexPoolSnapshot{
		ClipReserve: pool.ClipReserve, USDTReserve: pool.USDTReserve,
		ClipPerUSDT:        float64(pool.ClipReserve) / float64(pool.USDTReserve),
		USDTPerCLIP:        float64(pool.USDTReserve) / float64(pool.ClipReserve),
		TotalLiquidityUSDT: pool.USDTReserve * 2, UpdatedAt: pool.UpdatedAt,
	}, nil
}
